src/UrlSearchParams.ts

```diff
@@ -0,0 +1,151 @@
+import { duplicate } from './lang';
+
+/**
+ * A object with string keys and string or string array values that describes a query string.
+ */
+type ParamList = { [key: string]: string | string[] };
+
+/**
+ * Parse a query string, returning a ParamList object.
+ */
+function parseQueryString(input: string): ParamList {
+	const query = <{ [key: string]: string[] }> {};
+	input.split('&').forEach(entry => {
+		let [ key, value ] = entry.split('=');
+		key = key ? decodeURIComponent(key) : '';
+		value = value ? decodeURIComponent(value) : '';
+
+		if (key in query) {
+			query[key].push(value);
+		}
+		else {
+			query[key] = [ value ];
+		}
+	});
+	return query;
+}
+
+/**
+ * An class that represents a URL query search parameters.
+ */
+export default class UrlSearchParams {
+	/**
+	 * Construct a new UrlSearchParams from a query string, an object of parameters and values, or anoter
+	 * UrlSearchParams.
+	 */
+	constructor(input?: string | ParamList | UrlSearchParams) {
+		let list: ParamList;
+
+		if (input instanceof UrlSearchParams) {
+			// Copy the incoming UrlSearchParam's internal list
+			list = <ParamList> duplicate(input.list);
+		}
+		else if (typeof input === 'object') {
+			// Copy the incoming object, assuming its property values are either arrays or strings
+			list = {};
+			for (var key in input) {
+				const value = (<ParamList> input)[key];
+
+				if (Array.isArray(value)) {
+					list[key] = value.length ? value.slice() : [ '' ];
+				}
+				else if (value == null) {
+					list[key] = [ '' ];
+				}
+				else {
+					list[key] = [ <string> value ];
+				}
+			}
+		}
+		else if (typeof input === 'string') {
+			// Parse the incoming string as a query string
+			list = parseQueryString(input);
+		}
+		else {
+			list = {};
+		}
+
+		Object.defineProperty(this, 'list', { value: list });
```

Comment:
Is this because list is private?  I'm not sure there's much point to it being non-enumerable.  @BryanForbes, want to weigh in?